tachyon/crypto/commitments/fri/two_adic_fri_verifier.h

```diff
@@ -0,0 +1,133 @@
+                          evals[0]);
+    index = index_pair;
+  }
+  CHECK_LT(index, config.Blowup()) << "index was " << index;
+  CHECK_EQ(ro_idx, ro_size);
+
+  return folded_eval;
+}
+
+template <typename F, typename InputMMCS, typename ChallengeMMCS,
+          typename Challenger, typename ExtF = typename ChallengeMMCS::Field,
+          typename Function = base::RepeatingCallback<
+              void(size_t, std::vector<BatchOpening<InputMMCS>>,
+                   std::vector<size_t>, std::vector<ExtF>)>>
+[[nodiscard]] bool TwoAdicFriPCSVerify(
+    const TwoAdicFriConfig<ChallengeMMCS>& config,
+    const TwoAdicFriProof<ChallengeMMCS, std::vector<BatchOpening<InputMMCS>>,
+                          F>& proof,
+    Challenger& challenger, Function OpenInput) {
+  using Commitment = typename ChallengeMMCS::Commitment;
+  size_t num_commits = proof.commit_phase_commits.size();
+  std::vector<ExtF> betas = base::Map(
+      proof.commit_phase_commits,
+      [&challenger](size_t i, const Commitment& commit) {
+        challenger.ObserveContainer(commit);
+        ExtF beta = challenger.template SampleExtElement<ExtF>();
+        VLOG(2) << "FRI(beta[" << i << "]): " << beta.ToHexString(true);
+        return beta;
+      });
+  challenger.ObserveContainer(proof.final_eval);
+
+  if (proof.query_proofs.size() != config.num_queries) {
+    LOG(ERROR) << "proof size doesn't match " << proof.query_proofs.size()
+               << " vs " << config.num_queries;
+    return false;
+  }
+  // Check PoW.
+  VLOG(2) << "FRI(pow): " << proof.pow_witness.ToHexString(true);
+  if (!(challenger.CheckWitness(config.proof_of_work_bits,
+                                proof.pow_witness))) {
+    LOG(ERROR) << "failed to check pow";
+    return false;
+  }
+
+  uint32_t log_max_num_rows = num_commits + config.log_blowup;
+  std::vector<size_t> ro_num_rows;
+  std::vector<ExtF> ro_value;
+  std::vector<CommitStep<ChallengeMMCS>> steps(num_commits);
+
+  VLOG(2) << "FRI(final_eval): " << proof.final_eval.ToHexString(true);
+  for (size_t i = 0; i < proof.query_proofs.size(); ++i) {
+    ro_num_rows.clear();
+    ro_value.clear();
+    size_t index = challenger.SampleBits(log_max_num_rows);
+    VLOG(2) << "FRI(index[" << i << "]): " << index;
+    OpenInput(index, proof.query_proofs[i].input_proof, ro_num_rows, ro_value);
+
+#if DCHECK_IS_ON()
+    // Check reduced openings sorted by |num_rows| descending
+    DCHECK(base::ranges::is_sorted(ro_num_rows.begin(), ro_num_rows.end(),
+                                   base::ranges::greater()));
+#endif
+    for (size_t j = 0; j < num_commits; ++j) {
+      steps[j] = CommitStep<ChallengeMMCS>{
+          betas[j], proof.commit_phase_commits[j],
+          proof.query_proofs[i].commit_phase_openings[j]};
+    }
+    ExtF folded_eval = VerifyQuery(index, log_max_num_rows, config, steps,
+                                   std::move(ro_num_rows), std::move(ro_value));
```

Comment:
This function accepts const reference! 

```suggestion
                                   ro_num_rows, ro_value);
```